pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -374,46 +373,42 @@ private void doFlush(){
     /**
      * Release resources and cancel pending tasks.
      */
-    @Override
-    public void close() {
+    public CompletableFuture<Void> close() {
         // If batch feature is disabled, there is nothing to close, so set the stat only.
         if (!batchEnabled) {
             STATE_UPDATER.compareAndSet(this, State.OPEN, State.CLOSED);
-            return;
+            return CompletableFuture.completedFuture(null);
         }
         // If other thread already called "close()", so do nothing.
         if (!STATE_UPDATER.compareAndSet(this, State.OPEN, State.CLOSING)){
-            return;
+            return CompletableFuture.completedFuture(null);
         }
+        CompletableFuture closeFuture = new CompletableFuture();
         // Cancel pending tasks and release resources.
         singleThreadExecutorForWrite.execute(() -> {
             try {
                 if (state == State.CLOSED) {
+                    closeFuture.complete(null);
                     return;
                 }
                 // If some requests are flushed, BK will trigger these callbacks, and the remaining requests in should
                 // fail.
-                failureCallbackByContextAndRecycle(flushContext, BUFFERED_WRITER_CLOSED_EXCEPTION);
+                failureCallbackByContextAndRecycle(flushContext,
```

Comment:
Now `MlTransactionLogImpl` and `MlPendingAckStore` only have special treatment for `ManagedLedgerException`, If we create a new Exception, we also should change `MlTransactionLogImpl` and `MlPendingAckStore` for adaptation.Since this PR is already too large, I will create a new Exception in the following PR to solve it(same as https://github.com/apache/pulsar/pull/16758#discussion_r945521841). Thanks

Reply:
Sounds good